packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1548,3 +1569,93 @@ String nextPlatform(String currentPlatform, FeatureFlags featureFlags) {
       return 'android';
   }
 }
+
+/// The source code generator runs additional  tasks that produce Dart source
+/// files.
+class SourceCodeGenerator {
+  SourceCodeGenerator({
+    @required FileSystem fileSystem,
+    @required Logger logger,
+    @required Artifacts artifacts,
+    @required Platform platform,
+    @required ProcessManager processManager,
+  }) : _fileSystem = fileSystem,
+       _logger = logger,
+       _artifacts = artifacts,
+       _processManager = processManager,
+       _platform = platform;
+
+  static const String _kLocalizationDependency = 'l10n.yaml';
+
+  final FileSystem _fileSystem;
+  final Logger _logger;
+  final Platform _platform;
+  final Artifacts _artifacts;
+  final ProcessManager _processManager;
+
+  bool _generateLocalizations = true;
+  DateTime _lastCompiled;
+  Depfile _depfile;
+
+  /// Run code generators.
+  Future<void> generate() async {
+    if (_generateLocalizations) {
+      await _runCodeGenerators();
+    }
+  }
+
+  Future<void> _runCodeGenerators() async {
+    final File configFile = _fileSystem.file(_kLocalizationDependency);
+    _generateLocalizations = configFile.existsSync();
+    if (!_generateLocalizations) {
+      return;
+    }
+
+    final List<Uri> urisToScan = <Uri>[
+      configFile.uri,
+      if (_depfile != null)
+        ..._depfile.inputs.map((File file) => file.uri),
+      if (_depfile != null)
+        ..._depfile.outputs.map((File file) => file.uri),
+    ];
+
+    // Short-circuit as soon as a changed value is detected.
+    bool dirty = false;
+    if (_lastCompiled != null) {
+      for (final Uri uri in urisToScan) {
+        final DateTime updatedAt = _fileSystem.statSync(
+          uri.toFilePath(windows: _platform.isWindows)).modified;
+        if (updatedAt != null && updatedAt.isAfter(_lastCompiled)) {
+          dirty = true;
+          break;
+        }
+      }
+    } else {
+      dirty = true;
+    }
+    // If the depfile doesn't exist, we need to run at least once.
+    if (dirty || _depfile == null) {
+      final LocalizationOptions options = parseLocalizationsOptions(
+        file: configFile,
+        logger: _logger,
+      );
+      _logger.printTrace('Running update localizations');
+      try {
+        _depfile = await generateLocalizations(
```

Comment:
This is breaking the abstraction of the build system. I would have expected the ResidentRunner's code generation command to invoke source-generating build targets, and not reach directly into the build targets' implementation.

Reply:
I've thought about it for a bit and I don't really have a solution yet. I don't want to pull the entire build system into the resident runners, as they're already complicated. I also don't file hashes or file hash persistence. I think I need to set up support for a different kind of incremental build API.

The localizations are a good use case that would like me design something that works well, which I can either try to do before this PR lands or as a follow up. I don't have an estimate of how long that will take.

Reply:
I have some ideas I commented on the main thread

Reply:
see https://github.com/flutter/flutter/pull/56067